lib/segment/src/vector_storage/simple_multi_dense_vector_storage.rs

```diff
@@ -0,0 +1,124 @@
+    dim: usize,
+    distance: Distance,
+    stopped: &AtomicBool,
+) -> OperationResult<Arc<AtomicRefCell<VectorStorageEnum>>> {
+    let (mut deleted, mut deleted_count) = (BitVec::new(), 0);
+    let db_wrapper = DatabaseColumnWrapper::new(database, database_column_name);
+
+    let mut total_vector_count = 0;
+    let mut total_sparse_size = 0;
+    db_wrapper.lock_db().iter()?;
+    for (key, value) in db_wrapper.lock_db().iter()? {
+        let point_id: PointOffsetType = bincode::deserialize(&key)
+            .map_err(|_| OperationError::service_error("cannot deserialize point id from db"))?;
+        let stored_record: StoredMultiDenseVector = bincode::deserialize(&value)
+            .map_err(|_| OperationError::service_error("cannot deserialize record from db"))?;
+
+        // Propagate deleted flag
+        if stored_record.deleted {
+            bitvec_set_deleted(&mut deleted, point_id, true);
+            deleted_count += 1;
+        }
+        total_vector_count = std::cmp::max(total_vector_count, point_id as usize + 1);
+
+        check_process_stopped(stopped)?;
+    }
+
+    Ok(Arc::new(AtomicRefCell::new(
+        VectorStorageEnum::MultiDenseSimple(SimpleMultiDenseVectorStorage {
+            dim,
+            distance,
+            db_wrapper,
+            update_buffer: StoredMultiDenseVector {
+                deleted: false,
+                vector: MultiDenseVector::default(),
+            },
+            deleted,
+            deleted_count,
+            total_vector_count,
+        }),
+    )))
+}
+
+impl SimpleMultiDenseVectorStorage {
+    /// Set deleted flag for given key. Returns previous deleted state.
+    #[inline]
+    #[allow(unused)]
+    fn set_deleted(&mut self, key: PointOffsetType, deleted: bool) -> bool {
+        if key as usize >= self.total_vector_count {
+            return false;
+        }
+        let was_deleted = bitvec_set_deleted(&mut self.deleted, key, deleted);
+        if was_deleted != deleted {
+            if !was_deleted {
+                self.deleted_count += 1;
+            } else {
+                self.deleted_count = self.deleted_count.saturating_sub(1);
+            }
+        }
+        was_deleted
+    }
+
+    #[allow(unused)]
+    fn update_stored(
+        &mut self,
+        key: PointOffsetType,
+        deleted: bool,
+        vector: Option<&MultiDenseVector>,
+    ) -> OperationResult<()> {
+        // Write vector state to buffer record
+        let record = &mut self.update_buffer;
+        record.deleted = deleted;
+        if let Some(vector) = vector {
+            record.vector = vector.clone();
+        }
+
+        // Store updated record
+        self.db_wrapper.put(
+            bincode::serialize(&key).unwrap(),
+            bincode::serialize(&record).unwrap(),
+        )?;
+
+        Ok(())
+    }
+}
+
+// TODO integrate MultiDenseVector to Vectors enum to enable this implementation
+// impl VectorStorage for SimpleMultiDenseVectorStorage
```

Comment:
@IvanPleshkov implementing this trait means introducing a new type in the Vector hierarchy.

We have seen that movie already, better do it separately :sweat_smile: